libbeat/common/jsontransform/jsonhelper.go

```diff
@@ -27,8 +27,15 @@ import (
 )
 
 // WriteJSONKeys writes the json keys to the given event based on the overwriteKeys option and the addErrKey
-func WriteJSONKeys(event *beat.Event, keys map[string]interface{}, overwriteKeys bool, addErrKey bool) {
+func WriteJSONKeys(event *beat.Event, keys map[string]interface{}, expandKeys, overwriteKeys, addErrKey bool) {
 	logger := logp.NewLogger("jsonhelper")
+	if expandKeys {
+		if err := ExpandFields(keys); err != nil {
```

Comment:
if we error here `keys` is in an unknown state. Do we need to clone keys before this call in order to keep it intact on error? Logging the original document would be needed for users to understand why things did go wrong.

Reply:
I was expecting the original document to show up under `message`, like when a JSON decoding error occurs. That doesn't happen though. Is there any reason why we shouldn't do that, for a consistent debugging experience, instead of logging?

Reply:
I forgot to respond to the first part:

> if we error here keys is in an unknown state. Do we need to clone keys before this call in order to keep it intact on error?

As long as we include the original input (in `message`), I don't see a need. I'm not intimately familiar with Filebeat though.

Reply:
The JSON decoder in the log input does not store the original raw line in the `message` field. One can configure a custom `message` field (which is extracted from the json document, not the original line), but by default `message` will not be set. 

Anyways, the new fields are merged into the event after expansion via `WriteJSONFields`. Neither the processor nor the json decoder in the log input reference to any fields in `keys` (or `keys` itself). This should make the operation safe. No need to clone.